clients/src/main/java/org/apache/kafka/common/serialization/ListSerializer.java

```diff
@@ -77,21 +87,39 @@ public void configure(Map<String, ?> configs, boolean isKey) {
         }
     }
 
+    private void serializeNullIndexList(final DataOutputStream out, List<Inner> data) throws IOException {
+        List<Integer> nullIndexList = IntStream.range(0, data.size())
+                .filter(i -> data.get(i) == null)
+                .boxed().collect(Collectors.toList());
+        out.writeInt(nullIndexList.size());
+        for (int i : nullIndexList) out.writeInt(i);
+    }
+
     @Override
     public byte[] serialize(String topic, List<Inner> data) {
         if (data == null) {
             return null;
         }
-        final int size = data.size();
         try (final ByteArrayOutputStream baos = new ByteArrayOutputStream();
              final DataOutputStream out = new DataOutputStream(baos)) {
+            out.writeByte(serStrategy.ordinal()); // write serialization strategy flag
+            if (serStrategy == SerializationStrategy.NULL_INDEX_LIST) {
+                serializeNullIndexList(out, data);
+            }
+            final int size = data.size();
             out.writeInt(size);
             for (Inner entry : data) {
-                final byte[] bytes = inner.serialize(topic, entry);
-                if (!isFixedLength) {
-                    out.writeInt(bytes.length);
+                if (entry == null) {
+                    if (serStrategy == SerializationStrategy.NEGATIVE_SIZE) {
+                        out.writeInt(Serdes.ListSerde.NEGATIVE_SIZE_VALUE);
+                    }
+                } else {
+                    final byte[] bytes = inner.serialize(topic, entry);
+                    if (!isFixedLength || serStrategy == SerializationStrategy.NEGATIVE_SIZE) {
+                        out.writeInt(bytes.length);
```

Comment:
This is what I was talking about in https://github.com/apache/kafka/pull/6592#issuecomment-606252503

Even if we are dealing with primitives, and a user chose `SerializationStrategy.NEGATIVE_SIZE`, we would have to encode each primitive's size in our payload.

Reply:
Hey, sorry that I'm jumping in here after there's been a long discussion which I missed, but I'm wondering why the serialization strategy would be configurable? IIUC the serialization strategy correctly, one of them basically means "constant-size data/primitive type, don't encode the size only length of list" while the other means "variable-size data, encode the size of each element only"

I assume this is to allow users to indicate that their data is constant size when its a non-primitize type, to avoid the need to encode this same size data -- that makes sense to me. But I think we can simplify the API a bit so we don't have to let users shoot themselves in the foot, as you said earlier 🙂 

How about: if it's a primitive type, and we can detect this (I think we should be able to), then we never encode the size info. If a user opts to do so, just log a warning and ignore it.

By the way, this might also be due to some earlier discussion I missed, but I find the names of the two SerializationStrategy enums super confusing. How about just `VARIABLE_SIZE` and `CONSTANT_SIZE`? Imo it's better to describe what the enum actually _means_ than how its implemented, you can read the code to understand the latter. But you shouldn't need to read the code to understand what a config means. Plus, this way we have flexibility to change the underlying implementation if we ever need to without also having to change the enum names which are now a public API

Reply:
Hey, no worries.

For:
> How about: if it's a primitive type, and we can detect this (I think we should be able to), then we never encode the size info. If a user opts to do so, just log a warning and ignore it.

I think I am already doing that in the constructors:
```
    public ListSerializer(Serializer<Inner> serializer) {
        this.inner = serializer;
        this.isFixedLength = serializer != null && fixedLengthSerializers.contains(serializer.getClass());
        this.serStrategy = this.isFixedLength ? SerializationStrategy.NULL_INDEX_LIST : SerializationStrategy.NEGATIVE_SIZE;
    }

    public ListSerializer(Serializer<Inner> serializer, SerializationStrategy serStrategy) {
        this(serializer);
        this.serStrategy = serStrategy;
    }
```

If a user doesn't pass `serStrategy` flag, we pick the best one for her based on passed serializer.
If a user passes her own `serStrategy` flag, we simply obey to it. 
However, we don't print any warning logs, since I assumed if the user passes the flag, then she probably knows what she is doing.

What do you think? I could add a warning log otherwise.

Reply:
As per flag names, totally agreed. Changing them to `VARIABLE_SIZE` and `CONSTANT_SIZE`.

Reply:
Awesome. Maybe I misunderstood this comment:
>Even if we are dealing with primitives, and a user chooses SerializationStrategy.NEGATIVE_SIZE, we would have to encode each primitive's size in our payload.

or maybe you just wrote that a while ago and it's out of date. Anyways what we're doing now sounds good, no reason to encode extra data even if the user selects this strategy for some reason. But I do think we should at least log a warning telling them they made a bad choice and it will be ignored. Most likely they just didn't understand what the parameter meant, and it's a good opportunity to enlighten them

Reply:
Hmmm, I thought you want to simply warn the user that the serialization strategy she chose is not optimal. But seems like you want to ignore the choice completely.

Then it doesn't make sense to expose this flag at all for the user to change. Me and @mjsax were discussing it earlier https://github.com/apache/kafka/pull/6592#issuecomment-606277356

Reply:
Ah, sorry if that wasn't clear. Yes I was proposing to ignore the choice if a user selects the `VARIABLE_SIZE` strategy with primitive type data. And to also log a warning in this case so at least we're not just silently ignoring it.

But I think you may have a point that perhaps we don't need to expose this flag at all. There seems to be no reason for a user to explicitly opt-in to the `VARIABLE_SIZE` strategy. Perhaps a better way of looking at this is to say that this strategy is the default, where the default will be overridden in two cases: data is a primitive/known type, or the data is a custom type that the user knows to be constant size and thus chooses to opt-in to the `CONSTANT_SIZE` strategy.

WDYT? We could simplify the API by making this a boolean parameter instead of having them choose a `SerializationStrategy` directly, something like `isConstantSize`.

Reply:
Hmmm, @mjsax reasoning was that in the future we could introduce **more** serialization strategies (https://github.com/apache/kafka/pull/6592#discussion_r370513438)

Reply:
My feeling is, don't over-optimize for the future. If/when we do want to add new serialization strategies it won't be that hard to pass a KIP that deprecates the current API in favor of whatever new one they decide on. And it won't be much work for users to migrate from the deprecated API. I'm all for future-proofness but imo it's better to start out with the simplest and best API for the current moment and then iterate on that, rather than try to address all possible eventualities with the very first set of changes. The only exception being cases where the overhead of migrating from the initial API to a new and improved one would be really high, either for the devs or for the user or both. But I don't think that applies here.

That's just my personal take. Maybe @mjsax would disagree, or maybe not. I'll try to ping him and see what he thinks now, since it's been a while since that last set of comments. Until then, what's your opinion here?

Reply:
Ok, in this case, I think the best course of action is to completely remove `SerializationStrategy` flag, and replace it with a simple boolean. Do not expose it to the user, and automatically choose the strategy based on the type of data.

If you agree, I'll go ahead and make the change.

Reply:
Just to clarify you mean don't expose this to the user at all, right? That sounds completely fine to me. If there are enough people trying to serialize lists of custom classes with all constant data size who want this optimization exposed for general use, then someone will request the feature and we can go back and add it in. Then we can debate what the API should look like at that time, and keep things simple for now. Personally I suspect the vast majority of non-primitive data types are not going to be constant size anyways.

Given the above, I think whether to track the strategy as an actual `SerializationStrategy` enum vs a boolean flag becomes a matter of code style and personal preference, since it's no longer exposed to the user. So it's up to you whether you find the enum or the flag to be more readable or clean